Refactor withStyles

behnammodi opened the pull request:
@kof I did but I need your help,
Items:
- Flow announcements 
- 2  Tests failed
`should render a number of composed style rules with withStyles API`
`allows nested ThemeProviders with custom namespace`

SUMMARY of Test:
`713 tests completed`
`12 tests skipped`
`2 tests failed`

kof commented:
One of the errors from tests is the missing `props.theming` handling 

```
  React-JSS: theming withStyles()
    when theming object returned from createTheming is provided to injectSheet options
      ✖ allows nested ThemeProviders with custom namespace
        Chrome 90.0.4430 (Mac OS X 10.15.7)
      Error: expected undefined to sort of equal { color: '#aaa' }
          at Assertion.101.Assertion.assert (webpack://jss/node_modules/expect.js/index.js:96:0 <- packages/react-jss/tests/withStylesTheming.js:197:13)
          at Assertion.101.Assertion.eql (webpack://jss/node_modules/expect.js/index.js:230:0 <- packages/react-jss/tests/withStylesTheming.js:331:10)
          at Context.<anonymous> (webpack://jss/packages/react-jss/tests/withStylesTheming.js:400:43 <- packages/react-jss/tests/withStylesTheming.js:20995:86)

```

behnammodi commented:
> One of the errors from tests is the missing `props.theming` handling
> 
> ```
>   React-JSS: theming withStyles()
>     when theming object returned from createTheming is provided to injectSheet options
>       ✖ allows nested ThemeProviders with custom namespace
>         Chrome 90.0.4430 (Mac OS X 10.15.7)
>       Error: expected undefined to sort of equal { color: '#aaa' }
>           at Assertion.101.Assertion.assert (webpack://jss/node_modules/expect.js/index.js:96:0 <- packages/react-jss/tests/withStylesTheming.js:197:13)
>           at Assertion.101.Assertion.eql (webpack://jss/node_modules/expect.js/index.js:230:0 <- packages/react-jss/tests/withStylesTheming.js:331:10)
>           at Context.<anonymous> (webpack://jss/packages/react-jss/tests/withStylesTheming.js:400:43 <- packages/react-jss/tests/withStylesTheming.js:20995:86)
> ```

Could you advise?

kof commented:
when props.theming is received you need to handle that and use it instead of the theming from context, see the test

kof commented:
Just looked at it more closely, I was wrong about props, we don't receive theming over props, only theme, so its in options. Your mistake was not using theming.context

behnammodi commented:
> Just looked at it more closely, I was wrong about props, we don't receive theming over props, only theme, so its in options. Your mistake was not using theming.context

Yeas, I add `theming.context` but it not work again

kof commented:
> Yeas, I add theming.context but it not work again

There is another bug somewhere, it passed the first 2 assertions

```
      expect(themeReceivedInComponentA).to.eql(themeA)
      expect(themeReceivedInComponentB).to.eql(themeB)
```

Creating a PR

kof commented:
https://github.com/cssinjs/jss/pull/1508

see last commit

behnammodi commented:
I fixed

kof commented:
I think you have permissions to commit directly into that new PR

behnammodi commented:
> I think you have permissions to commit directly into that new PR

I haven't permissions, but anyway I fixed on this PR

behnammodi commented:
> I think you have permissions to commit directly into that new PR

Also, I apply your changes

kof commented:
I changed the base of this PR so that I can commit somewhere and we can merge it. Please rebase this PR.

In the last commit I fixed the second mistake, createUseStyles() is accepting theming, not theme. This has fixed that  failing test now.

behnammodi commented:
> I changed the target branch of this PR so that I can commit somewhere and we can merge it. Please rebase this PR.
> 
> In the last commit I fixed the second mistake, createUseStyles() is accepting theming, not theme. This has fixed that failing test now.

I fixed too

behnammodi commented:
@kof We just have 1 error in test cases

`should render a number of composed style rules with withStyles API`

kof commented:
> @kof We just have 1 error in test cases

yeah, I am confident you can figure it out

behnammodi commented:
@kof Thanks man, I hope :)

behnammodi commented:
@kof We don't have any corresponding test case about `should render a number of composed style rules with styled API` for `useStyles` so I'm completely confusing. I find a piece of code that is not inside `createUseStyles` but we have it on previous `withStyles`:

```js
if (sheet && registry) {
  registry.add(sheet)
}
```

And I don't know we need it or no

kof commented:
@behnammodi not sure what you are confused about, from what I can see we have the register.add() call inside manageSheet()

behnammodi commented:
> @behnammodi not sure what you are confused about, from what I can see we have the register.add() call inside manageSheet()

Yes, you're right 👍🏻

kof commented:
@behnammodi  I am assuming you will let me know if you are completely stuck with this

behnammodi commented:
> @behnammodi I am assuming you will let me know if you are completely stuck with this

Yes, In the last test case  I'm completely stuck

kof commented:
ok, going to look into it

behnammodi commented:
> ok, going to look into it

Awesome 👍🏻

kof commented:
Fixed, please rebase

kof commented:
A bunch of type errors, otherwise all tests passed

behnammodi commented:
> Fixed, please rebase

Thanks man 👍🏻 🙏🏻

behnammodi commented:
@kof Hi man, I fixed temporarily `flow` issue, so you can review it again

behnammodi commented:
@kof sorry for late, I had very busy week 🙏🏻

kof commented on packages/react-jss/src/withStyles.js:
```diff
@@ -44,23 +35,24 @@ const createWithStyles: CreateWithStyles = <Theme>(styles, options = {}) => {
     )
 
+    const useStyles = createUseStyles(
+      (styles: any),
```
we shouldn't be using any type

behnammodi replied:
fixed

kof commented on packages/react-jss/src/withStyles.js:
```diff
@@ -44,23 +35,24 @@ const createWithStyles: CreateWithStyles = <Theme>(styles, options = {}) => {
+        index,
+        name: displayName,
+        ...sheetOptions
+      }: any)
```
any

behnammodi replied:
Fixed, I found a way from your codes

kof commented on packages/react-jss/src/withStyles.js:
```diff
@@ -44,23 +35,24 @@ const createWithStyles: CreateWithStyles = <Theme>(styles, options = {}) => {
 
-      if (injectTheme && newProps.theme == null) newProps.theme = theme
+      if (injectTheme && newProps.theme == null) {
+        ;(newProps: any).theme = theme
```
- any
- I think we can avoid needing semicolon first

behnammodi replied:
Yes, but semicolon added by prettier and I don't know way

behnammodi replied:
I fixed

kof commented on packages/react-jss/src/withStyles.js:
```diff
@@ -70,8 +62,11 @@ const createWithStyles: CreateWithStyles = <Theme>(styles, options = {}) => {
 
     WithStyles.displayName = `WithStyles(${displayName})`
+
+    // $FlowFixMe[prop-missing]
```
not sure if we can but maybe there is a way to avoid this?

behnammodi replied:
I could not find

kof commented:
merged, thank you!

kof commented:
released in https://github.com/cssinjs/jss/releases/tag/v10.7.0
